x-pack/plugins/fleet/server/services/epm/kibana/assets/install.ts

```diff
@@ -79,29 +69,44 @@ export function createSavedObjectKibanaAsset(asset: ArchiveAsset): SavedObjectTo
   };
 }
 
-// TODO: make it an exhaustive list
-// e.g. switch statement with cases for each enum key returning `never` for default case
 export async function installKibanaAssets(options: {
   savedObjectsClient: SavedObjectsClientContract;
   pkgName: string;
   kibanaAssets: Record<KibanaAssetType, ArchiveAsset[]>;
-}): Promise<SavedObject[]> {
+}): Promise<SavedObjectsImportSuccess[]> {
   const { savedObjectsClient, kibanaAssets } = options;
+  const assetsToInstall = Object.entries(kibanaAssets).flatMap(([assetType, assets]) => {
+    if (!validKibanaAssetTypes.has(assetType as KibanaAssetType)) {
+      return [];
+    }
 
-  // install the assets
-  const kibanaAssetTypes = Object.values(KibanaAssetType);
-  const installedAssets = await Promise.all(
-    kibanaAssetTypes.map((assetType) => {
-      if (kibanaAssets[assetType]) {
-        return AssetInstallers[assetType]({
-          savedObjectsClient,
-          kibanaAssets: kibanaAssets[assetType],
-        });
-      }
+    if (!assets.length) {
       return [];
-    })
-  );
-  return installedAssets.flat();
+    }
+
+    const assetFilter = AssetFilters[assetType];
+    if (assetFilter) {
+      return assetFilter(assets);
+    }
+
+    return assets;
+  });
+
+  if (!assetsToInstall.length) {
+    return [];
+  }
+
+  // As we use `import` to create our saved objects, we have to install
+  // their references (the index patterns) at the same time
+  // to prevent a reference error
+  const indexPatternSavedObjects = getIndexPatternSavedObjects() as ArchiveAsset[];
+
+  const installedAssets = await installKibanaSavedObjects({
+    savedObjectsClient,
+    kibanaAssets: [...indexPatternSavedObjects, ...assetsToInstall],
```

Comment:
I think we need to only add the index patterns if they don't already exist, so we likely need to do a get on those first to see if they exist. Otherwise, I believe we'll overwrite any customizations to the index patterns (such as runtime fields, field popularity, etc.).

I know we consider the objects installed by Fleet as managed and subject to overwrite at any time, but I think this case is different and high-value enough for us to try to support. Is there any reason this would be a challenge to implement?

Reply:
Unfortunately we have to add the index patterns at the same time as the package assets in order to avoid some missing functionality in the `import` method which is that it doesn't use `resolve` when checking references, so all refs must be imported with the assets. 

I thought this was "fine-ish" for now as it mimics our current behaviour of recreating them each time (even though we dont need to do this any more)